flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/committer/IcebergCommitter.java

```diff
@@ -0,0 +1,434 @@
+import org.apache.iceberg.Snapshot;
+import org.apache.iceberg.SnapshotUpdate;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.flink.TableLoader;
+import org.apache.iceberg.flink.sink.CommitSummary;
+import org.apache.iceberg.flink.sink.DeltaManifests;
+import org.apache.iceberg.flink.sink.DeltaManifestsSerializer;
+import org.apache.iceberg.flink.sink.FlinkManifestUtil;
+import org.apache.iceberg.flink.sink.IcebergFilesCommitterMetrics;
+import org.apache.iceberg.io.WriteResult;
+import org.apache.iceberg.relocated.com.google.common.annotations.VisibleForTesting;
+import org.apache.iceberg.relocated.com.google.common.base.MoreObjects;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.util.PropertyUtil;
+import org.apache.iceberg.util.ThreadPools;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class implements the Flink SinkV2 {@link Committer} interface to implement the Iceberg
+ * commits. The implementation builds on the following assumptions:
+ *
+ * <ul>
+ *   <li>There is a single {@link IcebergCommittable} for every checkpoint
+ *   <li>There is no late checkpoint - if checkpoint 'x' has received in one call, then after a
+ *       successful run only checkpoints &gt; x will arrive
+ *   <li>There is no other writer which would generate another commit to the same branch with the
+ *       same jobId-operatorId-checkpointId triplet
+ * </ul>
+ */
+public class IcebergCommitter implements Committer<IcebergCommittable>, Serializable {
+  private static final long serialVersionUID = 1L;
+  private static final String MAX_COMMITTED_CHECKPOINT_ID = "flink.max-committed-checkpoint-id";
+  private static final Logger LOG = LoggerFactory.getLogger(IcebergCommitter.class);
+  private static final byte[] EMPTY_MANIFEST_DATA = new byte[0];
+  private static final CommitRequest<IcebergCommittable> EMPTY_COMMIT_REQUEST =
+      new CommitRequest<IcebergCommittable>() {
+        @Override
+        public IcebergCommittable getCommittable() {
+          return new IcebergCommittable(EMPTY_MANIFEST_DATA, "jobId", "operatorId", -1L);
+        }
+
+        @Override
+        public int getNumberOfRetries() {
+          return 0;
+        }
+
+        @Override
+        public void signalFailedWithKnownReason(Throwable t) {}
+
+        @Override
+        public void signalFailedWithUnknownReason(Throwable t) {}
+
+        @Override
+        public void retryLater() {}
+
+        @Override
+        public void updateAndRetryLater(IcebergCommittable committable) {}
+
+        @Override
+        public void signalAlreadyCommitted() {}
+      };
+
+  @VisibleForTesting
+  static final String MAX_CONTINUOUS_EMPTY_COMMITS = "flink.max-continuous-empty-commits";
+
+  public static final long INITIAL_CHECKPOINT_ID = -1L;
```

Comment:
should we use Flink's `CheckpointStoreUtil#INVALID_CHECKPOINT_ID` instead?

`INITIAL` is also not accurate here. technically, initial checkpoint id is `1` as defined in Flink code
<img width="476" alt="image" src="https://github.com/apache/iceberg/assets/1545663/bd5037ec-1138-415f-bd65-a4e4be0159d0">

Reply:
@stevenzwu 
Existing code  (`IcebergFilesCommitter` from 1.17 and 1.18) is using this pattern 
```
  private static final long INITIAL_CHECKPOINT_ID = -1L;
```
as `IcebergCommitter` is a new addition for SinkV2, I will do the changes there, but I will not change existing behavior